core/genesis/file.go

```
@@ -0,0 +1,251 @@
+	dirCreateMode = fs.FileMode(0o700)
+)
+
+// NewFileGenesisSource returns a new GenesisSource for the provided
+// source directory and the provided module name where it is assumed
+// that the encoded json data of file.
+func NewFileGenesisSource(sourceDir, moduleName string) GenesisSource {
+	return &FileGenesisSource{sourceDir: filepath.Clean(sourceDir), moduleName: moduleName}
+}
+
+// OpenReader opens the source file reading from the given parameters,
+// and returns a ReadCloser.
+// It will try to open the file in order following by:
+// <sourceDir>/<module>/<field>.json
+// <sourceDir>/<module>.json
+// <sourceDir>/genesis.json
+func (f *FileGenesisSource) OpenReader(field string) (io.ReadCloser, error) {
+	// try reading genesis data from <sourceDir>/<module>/<field>.json
+	fName := fmt.Sprintf("%s.json", field)
+	fPath := filepath.Join(f.sourceDir, f.moduleName)
+
+	fp, err := os.Open(filepath.Clean(filepath.Join(fPath, fName)))
+	if err == nil {
+		return fp, nil
+	}
+
+	// if cannot find it, try reading from <sourceDir>/<module>.json
+	fName = fmt.Sprintf("%s.json", f.moduleName)
+	fp, err = os.Open(filepath.Clean(filepath.Join(f.sourceDir, fName)))
+	if err == nil {
+		return fp, nil
+	}
+
+	// else try reading <field> field from <sourceDir>/genesis.json
+	fPath = filepath.Join(f.sourceDir, "genesis.json")
+	fp, err = os.Open(filepath.Clean(fPath))
+	if err == nil {
+		return fp, nil
+	}
+
+	return nil, fmt.Errorf("fail to open reader from %s: %v", fPath, err)
+}
+
+// ReadMessage is a unsupported op
+func (f *FileGenesisSource) ReadMessage(proto.Message) error {
+	return errors.New("unsupported op")
+}
+
+// ReadRawJSON returns a json.RawMessage read from the source file given by the
+// source directory and the module name. If the module name is empty, it will reads
+// raw JSON data from <sourceDir>/genesis.json
+func (f *FileGenesisSource) ReadRawJSON() (rawBz json.RawMessage, rerr error) {
+	if len(f.moduleName) == 0 {
+		f.moduleName = "genesis"
```

we should return an error here and not default to `genesis`